sw/device/lib/testing/test_coverage.c

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/uart.h"
+#include "sw/vendor/llvm_clang_rt_profile/compiler-rt/lib/profile/InstrProfiling.h"
+#include <stdint.h>
+
+/**
+ * When the linker finds a definition of this symbol, it knows to skip loading
+ * the object which contains the profiling runtime’s static initializer.
+ */
+int __llvm_profile_runtime;
+
+/**
+ * Start and end of `.init_array` section where the LLVM profiling runtime
+ * initialization function resides, defined in `flash_link.ld`.
+ */
+extern void (*_init_array_start[])(void);
+extern void (*_init_array_end[])(void);
+
+/**
+ * Computes the CRC32 of a buffer as expected by Python's `zlib.crc32()`. The
+ * implementation below is basically a simplified, i.e. byte-by-byte and without
+ * a lookup table, version of zlib's crc32. See
+ * https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/crc32.c,
+ * lines 111-112 and 276-279.
+ */
+static uint32_t crc32(char *buf, size_t len) {
+  // CRC32 polynomial.
+  static const uint32_t kPoly = 0xEDB88320;
+  // Since we use a contiguous buffer, we don't need to call this function
+  // multiple times. That's why `crc` is not a parameter and is initialized to
+  // zero.
+  uint32_t crc = 0;
+  crc = ~crc;
+  for (; len > 0; --len, ++buf) {
+    crc = crc ^ *buf;
+    for (uint8_t i = 0; i < 8; ++i) {
+      if (crc & 1) {
+        crc = kPoly ^ (crc >> 1);
+      } else {
+        crc = crc >> 1;
+      }
+    }
+  }
+  return ~crc;
+}
+
+/**
+ * Sends the given buffer as a hex string over UART.
+ */
+static void send_buffer(char *buf, size_t len) {
+  for (uint32_t i = 0; i < len; ++i) {
+    base_printf("%2X", buf[i]);
+  }
+}
```

Comment:
Maybe it's worth adding a non-standard specifier to base_printf that does this, in analogy to `%z`? That way you can use a single base_printf in function below.

Reply:
I'm also not sure if we would be able to handle length cleanly. @mcy?

Reply:
Our non-conforming `printf` provides a couple of non-standard format characters. `%z` consumes two arguments: a `size_t` and a `char*`, behaves like `%s` but with the given length. My suggestion was to add a format specifier (e.g., `%y` and `%Y`) which behave the same way, but which print hex rather than ASCII. We can punt this but it seems like such a useful formatting strategy that I thought adding it to print.h was worth it. *shrug*.

Reply:
Thanks for the explanation @mcy. I think we should definitely do this because the log library already has an efficient way for writing logs for DV. Let me do this in a separate PR once this lands to address the performance issue in DV. Updated #3155.